src/backend/distributed/master/master_repair_shards.c

```diff
@@ -238,45 +313,282 @@ RecreateTableDDLCommandList(Oid relationId)
 
 
 /*
- * CopyDataFromFinalizedPlacement copies a the data for a shard (identified by
- * a relation and shard identifier) from a healthy placement to one needing
- * repair. The unhealthy placement must already have an empty relation in place
- * to receive rows from the healthy placement. This function returns a boolean
- * indicating success or failure.
+ * EnsureShardCanBeRepaired checks if the given shard has a placement in the source node but
+ * not on the target node.
+ */
+static void
+EnsureShardCanBeRepaired(int64 shardId, char *sourceNodeName, int32 sourceNodePort,
+						 char *targetNodeName, int32 targetNodePort,
+						 bool unrestrictedCopy)
+{
+	List *shardPlacementList = ShardPlacementList(shardId);
+	ShardPlacement *sourcePlacement = NULL;
+	ShardPlacement *targetPlacement = NULL;
+
+	sourcePlacement = SearchShardPlacementInList(shardPlacementList, sourceNodeName,
+												 sourceNodePort);
+	if (sourcePlacement->shardState != FILE_FINALIZED)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("source placement must be in finalized state")));
+	}
+
+	/*
+	 * We do not check if table already exist in target node if we do unrestrictedCopy.
+	 * If unrestrictedCopy is true, that means we are probably in a replicate or
+	 * rebalance operation.
+	 */
+	if (unrestrictedCopy)
+	{
+		return;
+	}
+
+	targetPlacement = SearchShardPlacementInList(shardPlacementList, targetNodeName,
+												 targetNodePort);
+	if (unrestrictedCopy || targetPlacement->shardState != FILE_INACTIVE)
```

Comment:
I agree all three remarks;
-unnecessary check after returing
-naming
-upper level deciding about doRepair. (However this means we should add EnsureShardCanBeCopied along with EnsureShardCanBeRepaired and EnsureShardCanBeMoved)